### PR TITLE
SocketTransportOptions: don't own handle to avoid close side-effects.

### LINKED ISSUE #70311: Potential breaking change in socket close behavior in net6 over net5 on Linux
It appears that when a NET6 program closes in a non-graceful scenario such as SIG_QUIT, it will modify the SO_LINGER option on its open listening socket. Specifically, it sets SO_LINGER to true, with a timeout of 0 seconds. Kestrel is listening on that socket; we invoke KestrelServerOptions.ListenHandle(...) during startup, and the socket handle is passed in via environment variable - e.g. the listening socket lives longer than the server program.

The code that does that appears to be taking some sort of "abortive" path, as in
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L264

The code above seemed similar to us between NET5 and NET6, e.g. the behavior wouldn't change, unless the "abortive" had a different value in NET6 for shutdown initiated with SIG_QUIT.

We were able to prove that when the same dotnet program targeting NET5 is shut down using SIG_QUIT, it does not apply any modification to the listening socket, leading us to believe that the "abortive" path in the code above wasn't taken in NET5, but it's taken in NET6.

In our case, the listening socket was passed to the program at start-up time by handle. We expect our program to not modify attributes of this externally provided socket.

This change in behavior between NET5 and NET6 does not appear in the list of breaking changes for NET6 over NET5.
It appears to be a breaking change to us though. 

If this is indeed a breaking change, we'd like it to be documented.
Also, we need a way to prevent .NET runtime from modifying attributes of this socket.

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/70311.
Depends on .NET 7: https://github.com/dotnet/runtime/pull/72000.

@halter73 ptal.

### CODE REVIEW

File: src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs [hunk 1]
```diff
@@ -126,7 +126,7 @@ public static Socket CreateDefaultBoundListenSocket(EndPoint endpoint)
                 // the underlying socket is never closed and the transport manager can hang
                 // when it attempts to stop.
                 listenSocket = new Socket(
-                    new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: true)
+                    new SafeSocketHandle((IntPtr)fileHandle.FileHandle, ownsHandle: false)
```
Comment (campersau) [inline src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:129]: Maybe also update comment above?

Comment (tmds) [reply src/Servers/Kestrel/Transport.Sockets/src/SocketTransportOptions.cs:129]: Thanks, I've updated the comment.

**Review by halter73 — APPROVED** [review]
Thanks! We already have a regression test verifying we stop the accept loop on shutdown in this case, right?

**Comment by halter73** [comment]
[CanListenToOpenTcpSocketHandle](https://dev.azure.com/dnceng-public/public/_build/results?buildId=8613&view=ms.vss-test-web.build-test-results-tab&runId=171272&resultId=109524&paneView=debug) passed in under 2 seconds, so I think we're good.

**Comment by halter73** [comment]
@tmds do you think this is worth backporting for .NET 7? Changes targeting `main` right now aren't going to make it in until .NET 8.

+ @romanpolunin I think you original reported this issue. Is there a good workaround for Kestrel closing the handle? Is this blocking any migrations or anything like that?

**Comment by tmds** [comment]
@halter73 depending the OS, accepted sockets inherits some properties of the listen socket. I don't know if this is the case for `SO_LINGER` on Linux. The man page doesn't mention, and I haven't tried.
Maybe @romanpolunin can tell us what issues it caused and we can decide based on that.

**Comment by romanpolunin** [comment]
For the workaround question, @halter73 : our workaround was (and still is) to manually set the socket attributes when new process starts, in the code of that process. E.g. we anticipate the socket to have been corrupted by previous shutdown (if any), and just force the proper config back on it. I don't have that piece of source code in front of me as I'm writing this comment, but it's a trivial change of the linger option to what it is by default.

For the issues caused question, @tmds :  our ops folks did a thorough packet trace session on what was going on, and we basically proved that server was prematurely terminating the connection before client had a chance to receive sent data. The issue was visible on response sizes over certain (small) threshold, so one of our early workarounds was actually to reduce the size of the response so it'd fit into a single TCP frame. 

Eventually we settled on a workaround based on manual re-confguration of the socket upon process start.

**Pull request merged**